### PR TITLE
Replace uniswap v3 with uniswap v2

### PR DESCRIPTION
**Pull request by richmanbtc**
Changes

- Replaced uniswap v3 with uniswap v2.
- Modified some specification about fee and funding payment.
- Removed sqrtPriceLimitX96 option that is required for IOC order.
- Changed to use mark price for liquidation check instead of index price.
- Changed isOverPriceLimit rules.
- Changed order limit rules.
- Send a part of the liquidation fee to the insurance fund.
- use deps/perpdex-oracle-contract instead of @perp/perp-oracle-contract
- update Uniswap v2 pairCodeHash

todo (Out of scope for this PR)

1. deployment scripts
2. multi collateral
3. test code

## Uniswap V2 pairCodeHash

https://ethereum.stackexchange.com/questions/120767/calculates-the-create2-address-for-a-pair-without-making-any-external-calls

## Order limit rules

New rule pros

- same as FTX
- simple
- more decentralized (not require trusted backstop liquidity providers)

New rule

- Can't send any orders except liquidation if MF < MMF.

Old rule

- Can't send any orders except liquidation if MF < 0.
- Only backstop liquidity provider can liquidate if MF < 0. 

https://help.ftx.com/hc/en-us/articles/360024780511-Futures-Specs-Summary

## Send a part of the liquidation fee to the insurance fund

Pros

- Insurance fund fee implementation can be simplified.
- Easy to analyze insurance fund balance statistically.
- It's more like CEX mechanics than the original implementation.

## isOverPriceLimit

New rules pros

- simple.
- similar to CEX rules (order size limit, price bands in FTX).

New rules

- open/close/liquidation: _isOverPriceLimit after swap must be false.

Old rules

- open: _isOverPriceLimit and _isOverPriceLimit after swap must be false
- close: if _isOverPriceLimit or _isOverPriceLimit after swap is true, only partial position can be closed

## Use mark price for liquidation check

Changed to use mark price (market price) for liquidation check instead of index price.

Reasons

- Using TWAP for liquidation check is not common in CEX.
- Higher quality requirements for Oracle reduce portability.
- FTX uses market price as mark price.
- If there is a discrepancy between the price used for the liquidation check and the market price, the asset may be stolen.

## fee

- Changed to automatically reinvest like Uniswap V2.
- The fee is no longer managed separately from liquidity.
- Removed fee compensation logic that is needed for uniswap v3.

## funding payment

- Funding payment is now paid every 15 minutes.

## Diff

diff: https://github.com/perpdex/perpdex-contract/compare/4a826c72631f47dd68ae1444096674a74746986d...feature/builtin_amm

diff2: https://github.com/perpdex/perpdex-contract/compare/481737effc65e879c0ae49f100655504091e028d...feature/builtin_amm

## survey

- architecture: https://gist.github.com/richmanbtc/e7e0b540c6a630aff10bd133caaf7584
- vamm taker: https://gist.github.com/richmanbtc/0e682eaaa1c4c5a00bc967a9df09180c
- vamm maker: https://gist.github.com/richmanbtc/41a4e530d078e056aa063efd61c5e26d
- uniswap replace: https://gist.github.com/richmanbtc/f7b614a4a7bfdf35fb667aae3e7a1651
- uniswap replace2: https://gist.github.com/richmanbtc/677452bcff7d0a94389b8554f4f29694
- uniswap fee: https://gist.github.com/richmanbtc/dd976c3029a3727ef9fcf6ec71570c4e
- funding payment: https://gist.github.com/richmanbtc/396afd2f44a52d9cda46ac08dcd90cec

### PR Reminders

Be ware of the followings

- [ ] implement deployment script for your changes in deployment repo
- [ ] add verification in hardhat simulation(000-prepare-simulation-check, 902-newMarket-check or 903-simulation-check) and system test(901-system-test) in deployment repo
- [ ] update change log
 
---

1. **Contract**: make sure the code follows our convention, ex: naming, explicit returns, etc.; if uncertain, discuss with others

2. **Test**: make sure tests can cover most normal and edge cases; if not, open follow-up tickets. Also, look out for failed tests and fix them!

3. **CHANGELOG.md**: update when external interfaces are changed

4. Workflow: 
    - Github: assign the pr to yourself; if pairing, can merge directly; else, assign someone to help review
    - Asana: assign the corresponding ticket to yourself and leave the pr link on it for easier follow-ups
    - Discord: if someone is mentioned in this pr, tag on Discord

### CODE REVIEW

File: contracts/lib/UniswapV2Broker.sol [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+        uint256 price0Cumulative;
+        uint256 price1Cumulative;
+        if (IUniswapV2Pair(pair).price0CumulativeLast() == 0) {
+            priceCumulative = 0;
+            blockTimestamp = 0;
+        } else {
+            (price0Cumulative, price1Cumulative, blockTimestamp) = UniswapV2OracleLibrary.currentCumulativePrices(pair);
+            priceCumulative = IUniswapV2Pair(pair).token0() == baseToken ? price0Cumulative : price1Cumulative;
+        }
+    }
```
Comment (kei49) [inline contracts/lib/UniswapV2Broker.sol:199]: @richmanbtc 
I supposed these internal view functions needs to be called somewhere in the contracts, so that users(UI) can get access to them via some other external view functions of contracts.

Comment (richmanbtc) [reply contracts/lib/UniswapV2Broker.sol:199]: Thanks. I don't want to add unnecessary code and increase maintenance costs at the MVP stage, so implement it when it is really needed.
